psi4/driver/qcdb/gradparse.py

```diff
@@ -0,0 +1,94 @@
+#
+# @BEGIN LICENSE
+#
+# Psi4: an open-source quantum chemistry software package
+#
+# Copyright (c) 2007-2018 The Psi4 Developers.
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This file is part of Psi4.
+#
+# Psi4 is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# Psi4 is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+
+import numpy as np
+
+from .util import filter_comments
+
+# TODO: Include gradient loading feature, analagously to hessian loading.
+
+def to_string(grad, handle, dtype='psi4', mol=None, energy=None):
+    """Writes gradient in various formats.
+
+    Parameters
+    ----------
+    grad : ndarray
+        (nat, 3) gradient array.
+    handle : filename or file handle
+        If file handle, it must have been opened in binary, `wb`.
+    dtype : {'file11', 'GRD', 'minimal'}, optional
+        Format to write Hessian.
+            file11 is an old file format with the number of atoms, last energy and geometry, as well as gradient.
+            GRD is an ACESII file format with the number of atoms and geometry, as well as gradient.
+            minimal simply gives the number of atoms and the gradient.
+    mol : psi4.molecule or qcdb.molecule
+        The molecule the gradient is from. Needed only in file11 formatting.
+    energy : psi4.molecule or qcdb.molecule
+        The electronic energy of the molecule the gradient is evaluated from. Needed only in file11 format.
+
+    Returns
+    -------
+    None
+
+    """
+    # Validate input.
+    try:
+        assert grad.ndim == 2 and grad.shape[1] == 3
+    except AssertionError:
+        raise ValidationError("gradient shape much be number of atoms * 3. Received shape: {}".format(grad.shape))
+
+    if dtype in {'file11', 'GRD'}:
```

Comment:
square, not curly, brackets, I think.

also, default dtype is 'psi4', which then isn't an option.

Reply:
I'll change 'psi4' to 'file11' in the next commit, after tests pass. Curse my indecisiveness on naming the various gradient formats!

Curly braces is another way to initialize a set. Didn't this come up earlier in the review process, too?

Reply:
You're right -- I need to remember sets. But does this have an advantage over plain `if dtype in ['file11', 'GRD']:`?

Reply:
Practically, no. From how I learned Python, I was under the impression that while lists were usually taught first, lists were no more plain than sets were. If you think lists are the plainer/more-Pythonic type, I can change to a list.

Reply:
Mosty ordered vs unordered difference (doesn't matter here). For a medium number of items sets will be faster and to `log(N)` lookup. Here, sets may be slower due to overhead. Not entirely sure why, but this kind of operation is, seemingly, exclusively implemented as a list.